### PR TITLE
grpc: improve stream more reliable

### PR DESCRIPTION
**Pull request by Shikugawa**
On previous implementation, when old stream (called stream A) is in the process of inactivation, after server sent `RST_STREAM`, gRPC client should close stream A, then it will call `resetStream()` on stream A. It will be a potential bug. So we implement this inactivation process on gRPC client. With this approach, we never lose pending messages even if connection had disposed with any issue. Because when stream A disposed with pending messages, then they will be drained on client draining queue. After that stream B will be created with drained pending messages. It will process this 2 phase operation eternally.

### CODE REVIEW

**Comment by wu-sheng** [comment]
> Because when stream A disposed with pending messages, then they will be drained on client draining queue.

Is there an OOM precaution for the queue?

**Comment by Shikugawa** [comment]
@wu-sheng There is no limit on draining queue. So there is no OOM precaution.

**Comment by wu-sheng** [comment]
The context is, the OAP server may be not available in the long term.

**Comment by Shikugawa** [comment]
Is it means OAP server will restart periodically?

**Comment by wu-sheng** [comment]
> Is it means OAP server will restart periodically?

No, it doesn't. It should not. But in the real world, everything could happen, we can't afford the instrument app crashes because of this, right? Instrumentation library could not be the failure reason.

**Comment by Shikugawa** [comment]
> > Is it means OAP server will restart periodically?
> 
> No, it doesn't. It should not. But in the real world, everything could happen, we can't afford the instrument app crashes because of this, right? Instrumentation library could not be the failure reason.

Ah, got it. So to speak, we should prevent OOM kill. It will cause potentially these queues. right?

**Comment by wu-sheng** [comment]
> Ah, got it. So to speak, we should prevent OOM kill. It will cause potentially these queues. right?

Yes. We could abandon the traces if the OAP is not available, even stop tracing(we did this in Java agent).

**Comment by Shikugawa** [comment]
> > Ah, got it. So to speak, we should prevent OOM kill. It will cause potentially these queues. right?
> 
> Yes. We could abandon the traces if the OAP is not available, even stop tracing(we did this in Java agent).

In my consideration, we can achieve this just with ring buffer. But we can introduce healthcheck agent in the future, it requires periodical operation so we can use it after introducing event library like libuv.

File: test/buffer_test.cc [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+// limitations under the License.
+
+#include <gmock/gmock.h>
+#include <gtest/gtest.h>
+
+// #include "source/utils/circular_buffer.h"
+
+namespace cpp2sky {
+
+// class std::queueTest : public testing::Test {
```
Comment (wu-sheng) [inline test/buffer_test.cc]: Does this something forget to delete?

Comment (Shikugawa) [reply test/buffer_test.cc]: Yes. I forgot to uncomment this.

**Review by wu-sheng — APPROVED** [review]

**Pull request merged**